File: leetcode/131.cpp
class Solution {
public:
    bool is_palin(string &s, int l, int r) {
        if(l < 0 || r >= s.size() || l > r) return false;
        while(l < r) {
            if(s[l] != s[r]) return false;
            ++l;
            --r;
        }
        return true;
    }

    void helper(string &s, int index,
                vector<vector<bool>> &palin,
                vector<string> &buf,
                vector<vector<string>> &ans) {
        if(index == s.size()) {
            ans.push_back(buf);
            return ;
        }

        for(int j = index; j < s.size(); ++j) {
            if(palin[index][j]) {
                buf.push_back(s.substr(index, j-index+1));
                helper(s, j+1, palin, buf, ans);
                buf.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        int n = s.size();
        vector<vector<bool>> palin(n, vector<bool>(n, false));
        for(int i = 0; i < n; ++i) {
            for(int j = i; j < n; ++j) {
                palin[i][j] = is_palin(s, i, j);
            }
        }

        vector<string> buf;
        vector<vector<string>> ans;
        helper(s, 0, palin, buf, ans);
        return ans;
    }
};

File: leetcode/131.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def is_palin(ss: str) -> bool:
            l, r = 0, len(ss)-1
            while l < r:
                if ss[l] != ss[r]:
                    return False
                l, r = l+1, r-1
            return True

        def check(tokens: List[str]) -> bool:
            return all([is_palin(ss) for ss in tokens])

        ans: List[List[str]] = []
        for spec in range((1 << (len(s)-1))):
            buf: List[str] = [s[:1]]
            for i in range(len(s)-1):
                if (1 << i) & spec:
                    buf.append(s[i+1:i+2])
                else:
                    buf[-1] += s[i+1]
            if check(buf):
                ans.append(buf)
        return ans
